### ISSUE TITLE
`HsJoiner` should avoid copying messages when practical

### ISSUE THREAD

**Issue by briansmith** [rustls/rustls#668]
We have:
```
 self.buf
            .extend_from_slice(&payload.0[..]);
```

I'm guessing that the vast majority of the time `self.buf` is empty. In that case, instead of `extend_from_slice`, we should swap `self.buf` and `payload.0` to save an allocation and a copy. After all, almost no handshake messages are fragmented in real life, right? Maybe just certificate messages.

**Comment by briansmith** [comment]
I took a stab at this in PR #669.

**Comment by alcore** [comment]
Fairly radical question: is `HsJoiner` even necessary?

My reasoning is that:
During a handshake the messages are expected in a fixed order - the state handlers assert that. Additionally, no ApplicationData can be sent/received yet; getting an alert regardless of the order is effectively a termination and CCS messages also are expected in a fixed order. So potential interleaving is a non-issue because the handshake will just fail, unless I'm missing something.

During traffic I can _theoretically_ imagine that an implementation might send ApplicationData frames interleaved with the odd Handshake message (TLS1.3 tickets etc., key updates etc.), while _realistically_ I have never seen that in the wild under TCP. I understand QUIC or any other UDP transport will have different considerations, but I still suggest getting rid of `HsJoiner` entirely, because it doesn't (and it can't, really) tackle interleaved messages and merely works as a accumulator - but we've already got one accumulating buffer: the `Deframer`.

Under the current design, `HsJoiner` only gets to process full frames, so under the assumption that frames could arrive interleaved/out of order, due to https://github.com/ctz/rustls/blob/main/rustls/src/msgs/deframer.rs#L103 and `HsJoiner` not getting to take care of the message until it actually forms a full frame, there's no reason to maintain a separate buffer. Even if it did get all the relevant bytes pushed to it since byte 0 (frame::Type::Handshake) up until the end of the frame, so that other frames could arrive interleaved while this one was buffering, that `Vec<u8>` and the additional `VecDeque` are still not necessary.

Swapping https://github.com/ctz/rustls/blob/main/rustls/src/msgs/deframer.rs#L21 to a `Vec<u8>` and storing some additional state - some sort of a `handshake_frame_offsets: Option<(core::num::NonZeroUsize, core::num::NonZeroUsize)>`* gives us all we need to be able to simultaneously accumulate a full Handshake message plus one additional other type of frame:

- The first bytes in that buffer are _always_ a frame header. This means we _always_ know how many bytes we need to reserve in that buffer for the first message, regardless of whether that first frame is a Handshake or any other type (note that this suffers from the same unrealistic bug that currently exists, in that if we get interleaved messages with either not having a full frame header, the flow gets corrupted);
- When we get a frame::type::Handshake and `handshake_frame_offsets.is_some()`, we take `handshake_frame_offset.0` as the starting offset and `handshake_frame_offset.1` as the current write index. The initial buffer reservation happens when we get a frame header anyways, so we merely need to check we stay within those bounds.
- When we get frame::type::Handshake and `handshake_frame_offsets.is_none()`, we check whether we're already buffering any other frame and if so, respectively start writing the Handshake message past the other frame's reservation (again, we always know where since we know the length - we've got it buffered) and set `handshake_frame_offsets` accordingly;
- Respectively perform the checks and set offsets and reservations accordingly for all other types.

Essentially it'd be a sequential buffer with 2 'slots' so to speak, in most cases only the first one ever being utilized. It looks more convoluted than it is - I believe this boils down to a single branch (match), 2 at most. I'm aware this mixes concerns (the deframer would become a 'HandshakeMessageAccumulator' on top), but it gets rid of one entire processing layer in the hot path, one not-reusable-buffer and one queue (all of those living through the entire connection).

The swap from a `Box<[u8; Message::MAX_WIRE_SIZE]>` in the `Deframer` to a `Vec<u8>` also means we wouldn't be allocating 16KiB for each single connection (or wouldn't need to, at least - potentially allowing to make this configurable and/or request that buffer from the Application, later on return it for reuse - again, this lives for the entire lifetime). Most full handshakes are much smaller. I suspect we could get by with a default of 4KiB for most cases, while app data gets bounded by configuration. I suggest adding upper sanity bounds on the message payload sizes while at it, as there currently are none in place (Handshake messages can grow unbounded, I believe).

If this gets a go, I'd be happy to take a stab at it (no sooner than next weekend, though).

* (reducing the offset usize to u32 or even u16 would probably be sufficient but would require a change to the MTU configuration, I believe - bounding it to some sane numbers)
